### PR TITLE
Optimize representation of runfiles in compact execution log

### PR DESCRIPTION
**Pull request by fmeum**
Runfiles trees are now represented with a custom `RunfilesTree` message in the compact execution log. This allows using `InputSet`s to representing all artifacts staged at canonical locations, with only symlinks and root symlinks stored flattened and with explicit runfiles paths. 

Since runfile paths can collide, this change makes it necessary to preserve the order of elements in an `InputSet`. The previous representation as repeated ID fields for each type (file, symlink, directory) made this impossible, so the representation has been modified to reference all direct entry IDs in a single repeated field. Since this also reduces the potential for type mismatches between the ID field type and the referenced message type, all other typed IDs are replaced with untyped ID fields. By slightly tweaking the way IDs are generated for nested entries and not emitting IDs for entries that are never referenced (e.g. `Spawn`s), IDs are now consecutive, which simplifies the (possibly concurrent) bookkeeping for consumers by allowing them to use an array to store the entries.

Progress on #18643.

RELNOTES: The compact execution log now stores runfiles in a more compact representation that should reduce the memory overhead and log output size, in particular for test spawns. This change required breaking changes to the (experimental) log format.

### CODE REVIEW

File: src/main/protobuf/spawn.proto
Comment (fmeum) [inline src/main/protobuf/spawn.proto:261]: @tjgq I haven't fixed all tests and would also add more, but could you already take a quick look at the general idea, in particular the proto representation?

Comment (tjgq) [reply src/main/protobuf/spawn.proto:261]: The way you're going with the proto representation SGTM; I might have some terminology nits to pick later when I do a full review, but feel free to proceed.

Comment (fmeum) [reply src/main/protobuf/spawn.proto:261]: Existing tests should be fixed now, I also added more documentation. 

Since some tests may need to assert on the exact proto structure, could we do a first round of review before I add them? Happy to resolve nits. :-)

Comment (fmeum) [reply src/main/protobuf/spawn.proto:261]: Added one more comment to explain the precedence in case of path collisions. Please doublecheck it, this part is pretty subtle.

**Comment by fmeum** [comment]
I resolved the conflict and subsequently fixed the existing tests after the merge of the `SymlinkAction` PR.

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java:296]: Add an explicit case for `RUNFILES_TREE_ID` with a better error message. (Useful to anyone touching the compact log implementation in the future and using the reconstructor to verify that the changes are correct.)

File: src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java:477]: Please keep/adapt the javadoc.

File: src/main/java/com/google/devtools/build/lib/analysis/Runfiles.java
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/analysis/Runfiles.java:394]: `.runfile`, not `.runfiles`? (I didn't even know this existed until today.)

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java:250]: We could avoid a second traversal here by keeping track of a boolean during the first traversal, but up to you; we can defer until benchmarks prove it worthwhile.

**Review by tjgq — CHANGES_REQUESTED** [review]

**Comment by fmeum** [comment]
While adding more tests, I noticed that there is a pathological case that the new representation currently doesn't handle: If multiple artifacts at canonical locations collide, then the last in nested set order wins. Runfiles always use postorder, which I can almost emulate, but the distinction by type in `InputSet` loses some information about the order. For example, a directory might have come before a file, but `InputSet` doesn't allow me to see that.

@tjgq Since entry IDs are globally unique anyway, what do you think of merging `file_ids`, `directory_ids` and `unresolved_symlink_ids` into a single field that preserves the order?

**Comment by tjgq** [comment]
> While adding more tests, I noticed that there is a pathological case that the new representation currently doesn't handle: If multiple artifacts at canonical locations collide, then the last in nested set order wins. Runfiles always use postorder, which I can almost emulate, but the distinction by type in `InputSet` loses some information about the order. For example, a directory might have come before a file, but `InputSet` doesn't allow me to see that.
> 
> @tjgq Since entry IDs are globally unique anyway, what do you think of merging `file_ids`, `directory_ids` and `unresolved_symlink_ids` into a single field that preserves the order?

Oof, it's a disruptive change for any tools already consuming the compact format, but that's exactly why I wanted the format to remain experimental until Bazel 8 is released. Let's do it.

File: src/main/protobuf/spawn.proto
Comment (fmeum) [reply src/main/protobuf/spawn.proto:285]: I am leaning towards ditching this detailed description and instead point at the SpawnLogReconstructor and accompanying tests for reference. There are far too many edge cases (think `.runfile`) to describe in English in a proto comment.

I think that would be fine as I would expect consumers to either be interested in perfect reconstruction (in this case, they could use the reconstructor tool or faithfully port it to some other language) or in structural relations/diffing (in this case, they can just ignore all these subtleties and just diff the root relative paths of the artifacts and their contents).

What do you think?

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java
Comment (fmeum) [inline src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java:74]: This is very ugly, but I don't currently see a way around it. I hope that the namespace conflict between repo names and configs can be resolved before `--experimental_sibling_repository_layout` is flipped.

Note that in Bazel 9 and later, we could instead check for the presence of `+` in repo names and prohibit it in config mnemonics.

File: src/main/protobuf/spawn.proto
Comment (fmeum) [inline src/main/protobuf/spawn.proto:285]: I asked on Bazel slack and learned of some plans to maybe encode an entire Python venv into runfiles symlinks. That would certainly void the benefits of this PR. 

What do you think, should we optimize this case now or wait for this to become an issue in practice and then "fix" it by introducing a new field that encodes symlinks as a nested set of entries instead?

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogContext.java
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/exec/SpawnLogContext.java:130]: As I recall, the explicit `path` argument is required to make sure that the path was obtained from the action filesystem (usually a `RemoteActionFileSystem`), as otherwise `input.getPath()` will attempt to read from disk, which would fail when building without the bytes. Can you please verify that this is not an issue? (Note that `Path.isDirectory()` returns `false` for a nonexistent path and the vast majority of calls to `isInputDirectory` manage to avoid I/O, so the failure may be rare and silent.)

File: src/main/protobuf/spawn.proto
Comment (tjgq) [reply src/main/protobuf/spawn.proto:285]: Does this nested set already exist somewhere in Bazel memory (it doesn't look like it), or would this project also entail rearchitecting the internal representation of runfile trees? The one thing I'm concerned about is prematurely designing for a future that will never arrive, or that will arrive in a different form.

File: src/main/protobuf/spawn.proto
Comment (tjgq) [reply src/main/protobuf/spawn.proto]: Yeah, I concur that fully describing the algorithm in English feels hopeless. I'd be on board with leaving just a scary comment here, and redirecting people who sufficiently care to the reconstructor.

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java
Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java]: Yes, that's what I meant, sorry.

**Review by tjgq — CHANGES_REQUESTED** [review]
I still need to review most of this, but here are a few comments to (hopefully) move things along.

File: src/main/protobuf/spawn.proto
Comment (tjgq) [reply src/main/protobuf/spawn.proto:396]: Saving memory in the reconstructor is a great point - you've convinced me of the merit of (1). For (2) I'm okay with a monotonic implementation, I'd just rather not make a promise at this time (although it might very well not matter much due to Hyrum's Law).

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogContext.java
Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/exec/SpawnLogContext.java:130]: To be more specific, I think the only scenario where this matters is when an output declared as a file is materialized as a directory. For source directories, `isInputDirectory` should be able to exit early at the metadata check; and for file outputs materialized as files, `Path.isDirectory()` returns the correct result when the file is missing from disk.

My plan is to hardcode `--incompatible_disallow_unsound_directory_outputs` in Bazel 8 (which would render this case moot) but I'd still prefer this change to be submitted separately, in case I'm wrong and the `Path.isDirectory()` check turns out to matter in some other case (we've had a spatter of annoying bugs around this).

File: src/test/java/com/google/devtools/build/lib/exec/SpawnLogContextTestBase.java
Comment (tjgq) [inline src/test/java/com/google/devtools/build/lib/exec/SpawnLogContextTestBase.java:608]: Can we instead add a `createInputMetadataProvider` overload taking a `RunfilesTree` in addition to the list of artifacts and call it here? (Also for other test cases)

**Review by tjgq — CHANGES_REQUESTED** [review]
I've done a full review. I'll approve once the remaining comments are addressed.

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogContext.java
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/exec/SpawnLogContext.java:130]: Isn't this covered by `testFileOutputWithDirectoryContents`? I changed the logic to always go through `fileSystem` just in case.

File: src/main/protobuf/spawn.proto
Comment (fmeum) [inline src/main/protobuf/spawn.proto:276]: While implementing a consumer I noticed how painful it would be to support both versions if we were to reassign these IDs. I would thus lean towards reserving them. Do you consider that okay as well?

Comment (tjgq) [reply src/main/protobuf/spawn.proto:275]: For format detection, would it be better use the presence of one of the newly added fields in the `Invocation` message?

I thought reassigning might be better than reserving because, if you were to consume the new format using an old tool, reserving means depsets are silently (mis-)interpreted as empty, while reassigning makes it likely that the tool would crash, as some ids won't point to objects of the expected type. But I don't *super* care either way.

File: src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java:547]: I can't attach this review comment there, but can you amend the comment in line 148 (describing the purpose of entryMap) to also mention "runfiles tree"?

File: src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java:558]: Should this comment be moved to the reconstructor? The way I see it, the proto itself doesn't enforce a precedence, it's up to the reconstructor to visit fields in the correct order.

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java
Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java:78]: Yeah, I don't see any good alternatives either, and, since it currently looks like `--experimental_sibling_repository` layout won't ever be flipped, this might be forever :( Oh well.

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java:64]: N.B. if the `external` prefix ends up being configurable as it's currently being discussed at https://github.com/bazelbuild/bazel/issues/23576#issuecomment-2357558975, we'd also have to include it in the compact log. cc @lberki for awareness.

**Review by tjgq — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java:558]: It's actually outdated now that we don't preresolve the symlinks, so I just removed it. The `spawn.proto` file itself has a similar comment that's up to date.

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java:64]: Since `--experimental_sibling_repository_layout` will probably be removed at some point, should I just remove support for it here and then we can add a new field? I can also keep the support around for as long as the flag is there.

File: src/main/protobuf/spawn.proto
Comment (fmeum) [reply src/main/protobuf/spawn.proto:275]: > I suspect what you might be hinting at is that, with reserved fields, you can take a modified version of the proto with the removed fields added back, and use it to implement a single tool that understands both formats? But reassigning would require such a tool to interpret the proto contents at the byte level?

Yes, that's my point. I don't expect many external tools to consume the new format yet, but many users of such tools will be on Bazel < 7.4.0 for a while. If we want to make it easy for them to adopt the tools, I think that we should make it easy for tool authors to seamlessly support both.

Comment (tjgq) [reply src/main/protobuf/spawn.proto:275]: Understood, let's go with the current state.

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java
Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java:64]: Let's keep `--experimental_sibling_repository_layout` supported for now.

File: src/main/java/com/google/devtools/build/lib/exec/SpawnLogContext.java
Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/exec/SpawnLogContext.java]: Yes, it should have been covered. It's possible that it's no longer needed. I'll try to simplify in a followup; I just prefer not to risk a rollback of this very large PR.

**Review by tjgq — APPROVED** [review]
Thanks for this amazing contribution. I'll get started on the import.

**Comment by fmeum** [comment]
@bazel-io fork 7.4.0

**Comment by tjgq** [comment]
I'm making the following changes on import:

- Use new field ID for `InputSet.input_ids` and `Output.file_id` instead of reusing an existing ID (makes it easier to write a polyglot tool, as discussed out of band)
- Replace occurrences of `bazel` in tests with `TestConstants.PRODUCT_NAME` (which is `blaze` internally)

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!

**Pull request closed without merging**
